lib/gnat/services/server.ex

```diff
@@ -0,0 +1,167 @@
+  If an exception was raised during your `request/1` function, then the exception will be passed as the second argument.
+  If your `request/1` function returned something other than the supported return types, then its return value will be passed as the second argument.
+  """
+  @callback error(message::Gnat.message(), error::term()) :: :ok | {:reply, iodata()}
+
+  defmacro __using__(_opts) do
+    quote do
+      @behaviour Gnat.Services.Server
+
+      def error(_message, error) do
+        require Logger
+        Logger.error(
+          "Gnat.Server encountered an error while handling a request: #{inspect(error)}",
+          type: :gnat_server_error,
+          error: error
+        )
+      end
+
+      defoverridable error: 2
+    end
+  end
+
+
+  @typedoc """
+  Service configuration is provided as part of the consumer supervisor settings in the `service_definition` field.
+  You can specify _either_ the `subscription_topics` field for a reguar server or the `service_definition` field for a
+  new NATS service.
+
+  * `name` - The name of the service. Needs to conform to the rules for NATS service names
+  * `version` - A required version number (w/out "v" prefix) conforming to semver rules
+  * `queue_group` - An optional queue group for service subscriptions. If left off, "q" will be used.
+  * `description` - An optional description of the service
+  * `metadata` - An optional string->string map of service metadata
+  * `endpoints` - A required list of service endpoints. All services must have at least one endpoint
+  """
+  @type service_configuration :: %{
+    required(:name) =>         binary(),
+    required(:version) =>      binary(),
+    required(:endpoints) =>    [endpoint_configuration()],
+    optional(:queue_group) => binary(),
+    optional(:description) => binary(),
+    optional(:metadata) =>     map(),
+  }
+
+  @typedoc """
+  Each service configuration must contain at least one endpoint. Endpoints can manually specify their
+  subscription subjects or they can be derived from the endpoint name.
+
+  * `subject` - A specific subject for this endpoint to listen on. If this is not provided, then the endpoint name will be used.
+  * `name` - The required name of the endpoint
+  * `group_name` - An optional group to which this endpoint belongs
+  * `queue_group` - A queue group for this endpoint's subscription. If not supplied, "q" will be used.
+  * `metadata` - An optional string->string map containing metadata for this endpoint
+  """
+  @type endpoint_configuration :: %{
+    required(:name) => binary(),
+    optional(:subject) => binary(),
+    optional(:group_name) => binary(),
+    optional(:queue_group) => binary(),
+    optional(:metadata) => map()
+  }
+
+
+  @doc false
+  def execute(module, message, responder_pid) do
+    try do
+      {endpoint, group} = ServiceResponder.lookup_endpoint(responder_pid, message.topic)
+
+      case :timer.tc(fn -> apply(module, :request, [message, endpoint, group]) end) do
```

Comment:
I'm pretty sure `:timer.tc` provides an elapsed number of microseconds, not nanoseconds?

Reply:
yeah it looks like the default is microseconds, but you an pass a time-unit to get it back as nanoseconds if you want https://www.erlang.org/doc/man/timer#tc-1